controllers/reclaimspacecronjob_controller.go

```diff
@@ -0,0 +1,372 @@
+	if missedRun.IsZero() {
+		logger.Info("No upcoming scheduled times, sleeping until next")
+		return scheduledResult, nil
+	}
+
+	// make sure we're not too late to start the run
+	logger = logger.WithValues("currentRun", missedRun)
+	tooLate := false
+	if rsCronJob.Spec.StartingDeadlineSeconds != nil {
+		tooLate = missedRun.Add(time.Duration(*rsCronJob.Spec.StartingDeadlineSeconds) * time.Second).Before(r.Now())
+	}
+	if tooLate {
+		logger.Info("Missed starting deadline for last run, sleeping till next")
+		return scheduledResult, nil
+	}
+
+	// replace existing ones, if replace concurrent policy is set.
+	if rsCronJob.Spec.ConcurrencyPolicy == csiaddonsv1alpha1.ReplaceConcurrent {
+		err = r.Delete(ctx, activeJob, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "Failed to delete active reclaimSpaceJob", "reclaimSpaceJob", activeJob.Name)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// default is to forbid concurrent execution
+	if activeJob != nil {
+		logger.Info("Concurrency policy blocks concurrent runs, skipping", "activeJob", activeJob.Name)
+		return scheduledResult, nil
+	}
+
+	rsJob, err := r.constructRSJobForCronJob(rsCronJob, missedRun)
+	if err != nil {
+		logger.Error(err, "Failed to construct job from template")
+		// Do not requeue until spec is changed or till next scheduled time.
+		return scheduledResult, nil
+	}
+	if err := r.Create(ctx, rsJob); err != nil {
+		logger.Error(err, "Failed to create reclaimSpaceJob for reclaimSpaceCronJob")
+		return ctrl.Result{}, err
+	}
+	logger.Info("Successfully created reclaimSpaceJob for reclaimSpaceCronJob run", "reclaimSpacejob", rsJob.Name)
+
+	// Reconcile will be triggered if job starts or finishes, cronjob is modified, etc.
+	// Requeue till next run.
+	return scheduledResult, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ReclaimSpaceCronJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &csiaddonsv1alpha1.ReclaimSpaceJob{}, jobOwnerKey, func(rawObj client.Object) []string {
+		// extract the owner from job object.
+		job := rawObj.(*csiaddonsv1alpha1.ReclaimSpaceJob)
+		owner := metav1.GetControllerOf(job)
+		if owner == nil {
+			return nil
+		}
+		if owner.APIVersion != apiGVStr || owner.Kind != "ReclaimSpaceCronJob" {
+			return nil
+		}
+
+		return []string{owner.Name}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiaddonsv1alpha1.ReclaimSpaceCronJob{}).
+		Owns(&csiaddonsv1alpha1.ReclaimSpaceJob{}).
+		Complete(r)
+}
+
+// Now returns the current local time.
+func (r *ReclaimSpaceCronJobReconciler) Now() time.Time {
+	return time.Now()
+}
+
+// constructRSJobForCronJob constructs reclaimspacejob.
+func (r *ReclaimSpaceCronJobReconciler) constructRSJobForCronJob(
```

Comment:
keep it as a function not as a method as `r` is not used?

Reply:
```
	if err := ctrl.SetControllerReference(rsCronJob, job, r.Scheme); err != nil {
		return nil, err
	}
```
it is used.